hack/extract-release.sh: fails with Operation not permitted

Environment: vagrant

```
hack/extract-release.sh
tar: kube-controller-manager: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: kube-proxy: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: kube-scheduler: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: kubelet: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: oadm: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: oc: Cannot hard link to ‘kubectl’: Operation not permitted
tar: openshift: Cannot hard link to ‘kube-apiserver’: Operation not permitted
tar: Exiting with failure status due to previous errors
Makefile:145: recipe for target 'release' failed
make: *** [release] Error 2
```

Running with `sudo` didn't help.

--- a/hack/build-go.sh
+++ b/hack/build-go.sh
@@ -18,9 +18,9 @@ if [[ -z "$@" ]]; then
     platform=$(os::build::host_platform)
     echo "++ Using release artifacts from ${OS_RELEASE_COMMIT} for ${platform} instead of building"
     mkdir -p "${OS_OUTPUT_BINPATH}/${platform}"
-    tar mxzf "${OS_PRIMARY_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/${platform}"
-    tar mxzf "${OS_CLIENT_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/${platform}"
-    tar mxzf "${OS_IMAGE_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/${platform}"
+    os::build::extract_tar "${OS_PRIMARY_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/${platform}"
+    os::build::extract_tar "${OS_CLIENT_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/${platform}"
+    os::build::extract_tar "${OS_IMAGE_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/${platform}"
 
     os::build::make_openshift_binary_symlinks
 

--- a/hack/build-images.sh
+++ b/hack/build-images.sh
@@ -51,22 +51,33 @@ else
   imagedir="${OS_OUTPUT}/images"
   rm -rf ${imagedir}
   mkdir -p ${imagedir}
-  tar xzpf "${OS_PRIMARY_RELEASE_TAR}" --strip-components=1 -C "${imagedir}"
-  tar xzpf "${OS_IMAGE_RELEASE_TAR}" --strip-components=1 -C "${imagedir}"
+  os::build::extract_tar "${OS_PRIMARY_RELEASE_TAR}" "${imagedir}"
+  os::build::extract_tar "${OS_IMAGE_RELEASE_TAR}" "${imagedir}"
 fi
 
-# Copy primary binaries to the appropriate locations.
-ln -f "${imagedir}/openshift" images/origin/bin
-ln -f "${imagedir}/openshift" images/router/haproxy/bin
-ln -f "${imagedir}/openshift" images/ipfailover/keepalived/bin
+# Create link to file if the FS supports hardlinks, otherwise copy the file
+function ln_or_cp {
+  local src_file=$1
+  local dst_dir=$2
+  if os::build::is_hardlink_supported "${dst_dir}" ; then
+    ln -f "${src_file}" "${dst_dir}"
+  else
+    cp -pf "${src_file}" "${dst_dir}"
+  fi
+}
 
-# Copy image binaries to the appropriate locations.
-ln -f "${imagedir}/pod"             images/pod/bin
-ln -f "${imagedir}/hello-openshift" examples/hello-openshift/bin
-ln -f "${imagedir}/deployment"      examples/deployment/bin
-ln -f "${imagedir}/gitserver"       examples/gitserver/bin
-ln -f "${imagedir}/dockerregistry"  images/dockerregistry/bin
-ln -f "${imagedir}/recycle"         images/recycler/bin
+# Link or copy primary binaries to the appropriate locations.
+ln_or_cp "${imagedir}/openshift" images/origin/bin
+ln_or_cp "${imagedir}/openshift" images/router/haproxy/bin
+ln_or_cp "${imagedir}/openshift" images/ipfailover/keepalived/bin
+
+# Link or copy image binaries to the appropriate locations.
+ln_or_cp "${imagedir}/pod"             images/pod/bin
+ln_or_cp "${imagedir}/hello-openshift" examples/hello-openshift/bin
+ln_or_cp "${imagedir}/deployment"      examples/deployment/bin
+ln_or_cp "${imagedir}/gitserver"       examples/gitserver/bin
+ln_or_cp "${imagedir}/dockerregistry"  images/dockerregistry/bin
+ln_or_cp "${imagedir}/recycle"         images/recycler/bin
 
 # Copy SDN scripts into images/node
 os::provision::install-sdn "${OS_ROOT}" "${OS_ROOT}/images/node"

--- a/hack/common.sh
+++ b/hack/common.sh
@@ -446,6 +446,48 @@ os::build::archive_tar() {
   popd &>/dev/null
 }
 
+# Checks if the filesystem on a partition that the provided path points to is
+# supporting hard links.
+#
+# Input:
+#  $1 - the path where the hardlinks support test will be done.
+# Returns:
+#  0 - if hardlinks are supported
+#  non-zero - if hardlinks aren't supported
+os::build::is_hardlink_supported() {
+  local path="$1"
+  # Determine if FS supports hard links
+  local temp_file=$(mktemp --tmpdir="${path}")
+  ln "${temp_file}" "${temp_file}.link" &> /dev/null && unlink "${temp_file}.link" || local supported=$?
+  rm -f "${temp_file}"
+  return ${supported:-0}
+}
+
+# Extract a tar.gz compressed archive in a given directory. If the
+# archive contains hardlinks and the underlying filesystem is not
+# supporting hardlinks then the a hard dereference will be done.
+#
+# Input:
+#   $1 - path to archive file
+#   $2 - directory where the archive will be extracted
+os::build::extract_tar() {
+  local archive_file="$1"
+  local change_dir="$2"
+
+  # Unpack archive
+  echo "++ Extracting $(basename ${archive_file})"
+  if os::build::is_hardlink_supported "${change_dir}" ; then
+    tar mxzf "${archive_file}" --strip-components=1 -C "${change_dir}"
+  else
+    local temp_dir=$(mktemp -d --tmpdir=/dev/shm/)
+    tar mxzf "${archive_file}" --strip-components=1 -C "${temp_dir}"
+    pushd "${temp_dir}" &> /dev/null
+    tar cO --hard-dereference * | tar xf - -C "${change_dir}"
+    popd &>/dev/null
+    rm -rf "${temp_dir}"
+  fi
+}
+
 # os::build::make_openshift_binary_symlinks makes symlinks for the openshift
 # binary in _output/local/bin/${platform}
 os::build::make_openshift_binary_symlinks() {

--- a/hack/extract-release.sh
+++ b/hack/extract-release.sh
@@ -18,8 +18,8 @@ cd "${OS_ROOT}"
 os::build::detect_local_release_tars $(os::build::host_platform_friendly)
 
 mkdir -p "${OS_OUTPUT_BINPATH}/linux/amd64"
-tar mxzf "${OS_PRIMARY_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/linux/amd64"
-tar mxzf "${OS_CLIENT_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/linux/amd64"
-tar mxzf "${OS_IMAGE_RELEASE_TAR}" --strip-components=1 -C "${OS_OUTPUT_BINPATH}/linux/amd64"
+os::build::extract_tar "${OS_PRIMARY_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/linux/amd64"
+os::build::extract_tar "${OS_CLIENT_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/linux/amd64"
+os::build::extract_tar "${OS_IMAGE_RELEASE_TAR}" "${OS_OUTPUT_BINPATH}/linux/amd64"
 
 os::build::make_openshift_binary_symlinks

